sycl/plugins/opencl/pi_opencl.cpp

```diff
@@ -71,19 +73,92 @@ CONSTFIX char clGetDeviceFunctionPointerName[] =
 
 #undef CONSTFIX
 
+typedef CL_API_ENTRY cl_int(CL_API_CALL *clGetDeviceFunctionPointer_fn)(
+    cl_device_id device, cl_program program, const char *FuncName,
+    cl_ulong *ret_ptr);
+
+typedef CL_API_ENTRY cl_int(CL_API_CALL *clSetProgramSpecializationConstant_fn)(
+    cl_program program, cl_uint spec_id, size_t spec_size,
+    const void *spec_value);
+
+struct ExtFuncsPerContextT;
+
+namespace detail {
+template <const char *FuncName, typename FuncT>
+std::pair<FuncT &, bool &> get(ExtFuncsPerContextT &);
+} // namespace detail
+
+struct ExtFuncsPerContextT {
+#define _EXT_FUNCTION_INTEL(t_pfx)                                             \
+  t_pfx##INTEL_fn t_pfx##Func = nullptr;                                       \
+  bool t_pfx##Initialized = false;
+
+#define _EXT_FUNCTION(t_pfx)                                                   \
+  t_pfx##_fn t_pfx##Func = nullptr;                                            \
+  bool t_pfx##Initialized = false;
+
+#include "ext_functions.inc"
+
+#undef _EXT_FUNCTION
+#undef _EXT_FUNCTION_INTEL
+
+  std::mutex Mtx;
+
+  template <const char *FuncName, typename FuncT>
+  std::pair<FuncT &, bool &> get() {
+    return detail::get<FuncName, FuncT>(*this);
+  }
+};
+
+namespace detail {
+
+#define _EXT_FUNCTION_COMMON(t_pfx, t_pfx_suff)                                \
+  template <>                                                                  \
+  std::pair<t_pfx_suff##_fn &, bool &> get<t_pfx##Name, t_pfx_suff##_fn>(      \
+      ExtFuncsPerContextT & Funcs) {                                           \
+    using FPtrT = t_pfx_suff##_fn;                                             \
+    std::pair<FPtrT &, bool &> Ret{Funcs.t_pfx##Func,                          \
+                                   Funcs.t_pfx##Initialized};                  \
+    return Ret;                                                                \
+  }
+#define _EXT_FUNCTION_INTEL(t_pfx) _EXT_FUNCTION_COMMON(t_pfx, t_pfx##INTEL)
+#define _EXT_FUNCTION(t_pfx) _EXT_FUNCTION_COMMON(t_pfx, t_pfx)
+
+#include "ext_functions.inc"
+
+#undef _EXT_FUNCTION
+#undef _EXT_FUNCTION_INTEL
+#undef _EXT_FUNCTION_COMMON
+} // namespace detail
+
+struct ExtFuncsCachesT {
```

Comment:
Why plural form?
```suggestion
struct ExtFuncsCacheT {
```

Reply:
It's plural because each context has it's own cache.

Reply:
I think it would be reasonable to call the whole map a cache instead of ExtFuncsPerContextT instances.